### PR TITLE
storage: multi-worker persist_sink

### PR DESCRIPTION
**Pull request by guswynn**
This pr implements this design document: https://github.com/MaterializeInc/materialize/blob/9e688835d0f3ee646f9afd930dbd5b38d5a334cd/doc/developer/design/20230210_storage_persist_sink_redux.md

It also partially resolves https://github.com/MaterializeInc/materialize/issues/17221, but does not close it.

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

This branch is reviewable, but is stacked on top of https://github.com/MaterializeInc/materialize/pull/18010, which will need to be merged first. The structure of the branch is fairly complex:

- The first commit copies the compute `persist_sink` into `src/storage/src/render/multi_worker_persist_sink.rs`, but doesn't connect it to anything
- https://github.com/MaterializeInc/materialize/pull/17589/commits/5fff51f3f098e7f1cd92663d892bf9db6fbeb4eb actually connects the new implementation, behind a system parameter
  - Note that this system parameter is on-by-default in mzcompose's, because we want the new implementation to be the canonical implementation tested in ci. Unfortunately, the single-worker one will be used with `bin/environmentd` workflows until the default for the flag is changed to `true`
- The other commits iteratively change the compute implementation to match what is described in the design document!
 
### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
  - I will run nightlies and release-qualification tests on this, now published branch, but I tested the core implementation against them earlier in this pr's life, and they passed:
    - nightlies: https://buildkite.com/materialize/nightlies/builds/1796
    - release-qualification: https://buildkite.com/materialize/release-qualification/builds/87
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  - https://github.com/MaterializeInc/materialize/blob/9e688835d0f3ee646f9afd930dbd5b38d5a334cd/doc/developer/design/20230210_storage_persist_sink_redux.md
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

### CODE REVIEW

**Comment by danhhz** [comment]
View of the diffs past the first one that copies in the compute version of persist_sink: https://github.com/MaterializeInc/materialize/compare/fcbd2d34e7d0d9ebfc6d3f5153dbb340914da665..6867863a814a2ab6405f53bfff6d4f84c54cfedf

File: src/storage/src/render/persist_sink.rs
Comment (danhhz) [inline src/storage/src/render/persist_sink.rs:317]: can we move this check up before `open_writer`? that would save us a bunch of (linearized) writer registrations

File: src/storage/src/render/persist_sink.rs
Comment (danhhz) [inline src/storage/src/render/persist_sink.rs:329]: woah, quite impressed with rust's reachability checking here 🤯

**Comment by guswynn** [comment]
Thanks @danhhz !!

File: src/storage/src/render/persist_sink.rs
Comment (danhhz) [inline src/storage/src/render/persist_sink.rs:681]: don't think so, am I missing something?

File: src/storage/src/render/persist_sink.rs
Comment (danhhz) [inline src/storage/src/render/persist_sink.rs:837]: ditto here, can this be moved before the open_writer call?

**Review by danhhz — COMMENTED** [review]
cool!

(I think I found a couple potentially significant perf issues (the open_writer comments), but otherwise direction is very exciting!)

File: src/storage/src/render/persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/persist_sink.rs:681]: I think in the past it was...

Comment (danhhz) [reply src/storage/src/render/persist_sink.rs:681]: maybe it's the latter then. do you remember why?

**Comment by guswynn** [comment]
There seems to be 1 transient double-write issue: https://buildkite.com/materialize/tests/builds/49847#01864cc8-a464-4835-baa3-f8893a047db6 but its unclear if its actually a problem with the sink, or debezium messing up

cc @philip-stoev wondering if we have seen this before

**Comment by guswynn** [comment]
I will run the failing test a few times locally to see if I can re-create it

**Comment by guswynn** [comment]
```


> SELECT f1, f2, SUM(LENGTH(f3)) FROM postgres_source_tableC GROUP BY f1, f2;
--
  | rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 49s 74s 45s
  | 22:1: error: non-matching rows: expected:
  | [["A", "8", "102400"], ["B", "8", "102400"], ["C", "7", "102400"], ["D", "6", "102400"], ["E", "5", "102400"], ["F", "4", "102400"], ["G", "3", "102400"], ["H", "2", "102400"]]
  | got:
  | [["A", "8", "204800"], ["B", "8", "204800"], ["C", "7", "204800"], ["D", "6", "204800"], ["E", "5", "204800"], ["F", "4", "204800"], ["G", "3", "204800"], ["H", "2", "204800"]]
  | Poor diff:
  | - A 8 102400
  | + A 8 204800
  | - B 8 102400
  | + B 8 204800
  | - C 7 102400
  | + C 7 204800
  | - D 6 102400
  | + D 6 204800
  | - E 5 102400
  | + E 5 204800
  | - F 4 102400
  | + F 4 204800
  | - G 3 102400
  | + G 3 204800
  | - H 2 102400
  | + H 2 204800
  |  
  | \|
  | 21 \|
  | 22 \| > SELECT f1, f2, SUM(LENGTH(f3)) FROM postgres_source_tableC GROUP BY f1, f2;
  | \| ^


```

Is the error I saw, but again, its very unclear if its a debezium issue or not; I have not seen any other evidence that the sink is broken

File: src/storage/src/render/persist_sink.rs
Comment (antiguru) [inline src/storage/src/render/persist_sink.rs:335]: This unfortunately creates a copy of the data which is then discarded immediately. An alternative would be for this operator to have two outputs where one is the data and the other is the batch description. The underlying issue is that there are multiple consumers of a stream, which means that all but one receive a copy of the data. Can explain more if needed!

Comment (danhhz) [reply src/storage/src/render/persist_sink.rs:335]: Interesting! Makes sense. So you're saying instead of splitting the input stream to both operator 1 and 2, pipeline the data through op1 into op2 so op1 can just forward the data along? I just checked and it seems like the current compute persist_sink might be doing the bad thing here, can you confirm?

Comment (guswynn) [reply src/storage/src/render/persist_sink.rs:335]: this copy shouldn't be _too_ bad in the storage case, because the input is pipelined...

A second output is a good idea, but we need to be careful about output frontiers!

File: src/storage/src/internal_control.rs
Comment (antiguru) [inline src/storage/src/internal_control.rs:30]: Can we find a better name? `enable_multi_worker_persist_sink`?

Comment (guswynn) [reply src/storage/src/internal_control.rs:30]: yes!

Comment (guswynn) [reply src/storage/src/internal_control.rs:30]: Fixed in https://github.com/MaterializeInc/materialize/pull/18010

File: src/storage/src/render/persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/persist_sink.rs]: we need the `write.upper()` here

Comment (guswynn) [reply src/storage/src/render/persist_sink.rs]: oh wait on the false can I probably can!

File: src/sql/src/vars.rs
Comment (jkosh44) [inline src/sql/src/vars.rs:422]: We usually end the description of Materialize vars with `(Materialize).`, see above.

File: src/sql/src/vars.rs
Comment (jkosh44) [inline src/sql/src/vars.rs:1185]: Any specific reason to create a new category instead of adding it to `features`?

**Review by jkosh44 — APPROVED** [review]
surfaces changes LGTM modulo a couple small nits.

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (danhhz) [inline src/storage/src/render/multi_worker_persist_sink.rs:1006]: am I missing something or could this just do a for loop over batches to `+=` the metrics?

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (danhhz) [inline src/storage/src/render/multi_worker_persist_sink.rs:687]: why would it be?

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (danhhz) [inline src/storage/src/render/multi_worker_persist_sink.rs:576]: could we use something like `batch_descriptions_frontier` as a tighter lower bound?

**Review by danhhz — APPROVED** [review]
seems sane to me! I'd prefer that we also get a stamp from @aljoscha and/or @petrosagg before merging

**Comment by danhhz** [comment]
Oh, and here's what I was using to look at just the diffs from the copied in version (ty Gus for the commit structure, huge help!) https://github.com/MaterializeInc/materialize/compare/375da4df4e63af8db607951b36a1529849ae1e91...guswynn:storage-persist-sink-2#diff-9f55aca4c9ae9406033d40707e7066fc16f696a4cccc60b7f5bc05ac20340123R781

File: src/sql/src/vars.rs
Comment (guswynn) [reply src/sql/src/vars.rs:422]: Oh this is here because its stacked on top of this pr: https://github.com/MaterializeInc/materialize/pull/18010, so ill fix it there!

File: src/sql/src/vars.rs
Comment (guswynn) [reply src/sql/src/vars.rs:1185]: its not a sql feature!

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:781]: they can't! Ill add a comment

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:576]: I don't think we can, but i can't remember lucidly why; ill think about it overnight

**Comment by danhhz** [comment]
the more I've let this sit, the more I'm a little stressed about not including an equivalent to https://github.com/MaterializeInc/materialize/pull/17526 here. doesn't have to be this PR, but I think we need it before removing the ability to toggle back to the old storage persist_sink

**Comment by guswynn** [comment]
@danhhz 

> the more I've let this sit, the more I'm a little stressed about not including an equivalent to https://github.com/MaterializeInc/materialize/pull/17526 here. doesn't have to be this PR, but I think we need it before removing the ability to toggle back to the old storage persist_sink

Going to do this as a separate pr!

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:1006]: I need a vector of batches, without the batch metrics for the `compare_and _append`!

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs]: This was fairly complex, but I added a commit on top that does this! I will be running nightly and RQ's against it!

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:319]: Can we use `into_element` from `mz_ore` or otherwise assert that there is only a single capability in there?

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:330]: Can we use the resumption frontier from the configuration here to avoid this dance with persist?

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:365]: I was about to write the same comment. `Pipeline` doesn't save us, the operator structure in this sink will still clone every single container that passes through so I think we should actually have two outputs coming out of this operator, one for descriptions and one that simply forwards the data

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:360]: Can this not become a `while let Some(event) = desired_input.next().await ` ?

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:386]: We should use the `Antichain::as_option` function here because `.first()` is suspicious in non-total order times

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:389]: NIT: this will probably be much shorter with:

```rust
let Ok(cap) = cap_set.try_delayed(&batch_ts) else {
    panic!(
        "minter cannot delay {cap_set:?} to {lower:?}. Likely because \
        we already emitted a batch description and delayed."
    );
};
```

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:414]: Indeed, we should instead downgrade to the batch's `upper`

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:423]: We should just call `cap_set.downgrade()` if we don't have any additional context to add to the panic

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:434]: Can we move this inside the operator above?

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:450]: Could be a `&str` too

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:478]: If we assume that the data comes pre-sharded we should document that in the doc comments.

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:518]: In the rest of the codebase whenever `timely::progress::Timestamp` and `mz_repr::Timestamp` coexist we use `Timestamp` for the timely trait and `mz_repr::Timestamp` for the concrete mz timestamp. Can we follow this pattern here too?

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:564]: Why do we care about this frontier? Batch descriptions are guaranteed to not be overlapping and we can process them immediately

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:530]: Doing this will win you an indentation level:

```suggestion
                Some(event) = descriptions_input.next_mut() => match event {
```

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:549]: We should use the capability that came with the batch as-is, right?

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:548]: Here we are unconditionally cloning the description just to run an assertion. I suggest that we `assert!(!in_flight_batches.contains(&description), ...)` before instead.

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:551]: we should do this instead to avoid the unconditional clone of the description above:

```rust
if let Some(description) = existing {
    panic!(...);
}
```

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:587]: Where does the number 100 appear in the code? I don't see it

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:168]:
```suggestion
struct BatchBuilderAndMetadata<K, V, T, D>
```

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:689]: nit: redundant `.into_iter()` call

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:884]: We should leave a `TODO` here to add the correct schema

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:917]: I don't think this operator needs the description input, does it? Observing the batch stream and its frontiers should have enough data to do the compare and appends which would simplify this operator a lot

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [inline src/storage/src/render/multi_worker_persist_sink.rs:1027]: nit: redundant `.into_iter()` call

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:587]: oh sorry, this is an old comment

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:884]: This is the value schema, which is always empty

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:288]: I don't want to add a `since`, even if its a copy of a `lower`

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:389]: Im anti `let-else`!

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:564]: We use this as a light optimization on avoiding work after each `select!` to avoid emitting new batches if this or the desired frontier hasn't moved!

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:530]: oh wait, `select!` doesnt support this :(

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:551]: this became way more annoying because the description is the key, but the entry api worked

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:330]: The resumption frontier could be a minimum between many subsources, so I don't think so!

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:919]: This is how it works in compute, and this is the least-changed operator between them. I suspect we do it this way because it ensures our batches are grouped the same way they were emitted, which less a correctness requirement, and more a nice-to-have. @aljoscha might know more. This is another example of something that would best be fixed in a later pr, to both sinks, like https://github.com/MaterializeInc/materialize/pull/17589#discussion_r1106016139

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs]: Unfortunately, the borrowchecker has some problems here :(

**Comment by guswynn** [comment]
@petrosagg I believe I fixed or responded to all your comments, but I scheduled some time in the morning tomorrow to discuss, just in case!

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [reply src/storage/src/render/multi_worker_persist_sink.rs:388]: Spicy take. Still, a match is much sorter than the closures but no strong opinions

```rust
let cap = cap_set.try_delayed(&batch_ts) {
    Ok(cap) => cap,
    Err(_) => panic!(
        "minter cannot delay {cap_set:?} to {lower:?}. Likely because \
        we already emitted a batch description and delayed."
    ),
};
```

File: src/storage/src/render/multi_worker_persist_sink.rs
Comment (petrosagg) [reply src/storage/src/render/multi_worker_persist_sink.rs:520]: It does, we do it in a bunch of places. You just need to put a trailing comma because it's an expression like so: `match event { .. },`

Comment (guswynn) [reply src/storage/src/render/multi_worker_persist_sink.rs:520]: UGH the trailing comma, that must have been it

**Comment by guswynn** [comment]
@petrosagg I incorporated your change by cherry-picking and editing the commit!

I think everything else is resolved! https://github.com/MaterializeInc/materialize/pull/18241 will also be attached to this pr, before or after it merges, depending on when that pr is approved!

**Review by petrosagg — APPROVED** [review]

**Pull request merged**
